File: frontend/src/services/messageService.ts
import api from './api';
import type { ApiResponse } from '../types';

export interface InboxMessageItem {
  id: number;
  message_id: number;
  delivered_at: string;
  is_read: boolean;
  read_at: string | null;
  title: string;
  content: string;
  is_broadcast: boolean;
  created_at: string;
  expires_at: string | null;
  sender_username: string | null;
}

export interface InboxResponse {
  items: InboxMessageItem[];
  pagination: {
    page: number;
    pageSize: number;
    total: number;
    totalPages: number;
  };
}

export const messageService = {
  async getInbox(page = 1, pageSize = 10): Promise<InboxResponse> {
    const response = await api.get<ApiResponse<InboxResponse>>(`/messages/inbox?page=${page}&pageSize=${pageSize}`);
    if (response.data.success && response.data.data) {
      return response.data.data;
    }
    throw new Error(response.data.error?.message || 'Failed to fetch inbox');
  },

  async getUnreadCount(): Promise<number> {
    const response = await api.get<ApiResponse<{ unread: number }>>('/messages/unread-count');
    if (response.data.success && response.data.data) {
      return Number(response.data.data.unread || 0);
    }
    throw new Error(response.data.error?.message || 'Failed to fetch unread count');
  },

  async markRead(deliveryId: number): Promise<void> {
    const response = await api.post<ApiResponse<{ delivery: { id: number } }>>(`/messages/${deliveryId}/read`);
    if (!response.data.success) {
      throw new Error(response.data.error?.message || 'Failed to mark message as read');
    }
  },

  async markAllRead(): Promise<number> {
    const response = await api.post<ApiResponse<{ updated: number }>>('/messages/read-all');
    if (response.data.success && response.data.data) {
      return Number(response.data.data.updated || 0);
    }
    throw new Error(response.data.error?.message || 'Failed to mark all messages as read');
  },

  async sendByAdmin(payload: {
    title: string;
    content: string;
    is_broadcast: boolean;
    recipient_user_ids?: number[];
    expires_at?: string | null;
  }): Promise<{ delivery_count: number }> {
    const response = await api.post<ApiResponse<{ delivery_count: number }>>('/messages/admin/send', payload);
    if (response.data.success && response.data.data) {
      return { delivery_count: Number(response.data.data.delivery_count || 0) };
    }
    throw new Error(response.data.error?.message || 'Failed to send site message');
  },
};

File: client-desktop/src/HoYoMusic.Desktop.Infrastructure/Services/MessageService.cs
using System.Net.Http.Json;
using HoYoMusic.Desktop.Core.Abstractions;
using HoYoMusic.Desktop.Core.Contracts;
using HoYoMusic.Desktop.Core.Models;

namespace HoYoMusic.Desktop.Infrastructure.Services;

public sealed class MessageService : HoYoApiClient, IMessageService
{
    public MessageService(HttpClient httpClient, ITokenStore tokenStore) : base(httpClient, tokenStore) { }

    public async Task<InboxMessagesResponseData> GetInboxMessagesAsync(int page = 1, int pageSize = 20, CancellationToken ct = default)
    {
        var query = new QueryStringBuilder()
            .Add("page", Math.Max(1, page))
            .Add("pageSize", Math.Clamp(pageSize, 1, 100))
            .ToString();
        return await GetAuthedAsync<InboxMessagesResponseData>(HttpMethod.Get, $"messages/inbox{query}", "Failed to load inbox.", ct);
    }

    public async Task<int> GetUnreadCountAsync(CancellationToken ct = default)
    {
        var result = await GetAuthedAsync<UnreadCountResponseData>(HttpMethod.Get, "messages/unread-count", "Failed to load unread count.", ct);
        return result.Unread;
    }

    public async Task MarkMessageReadAsync(int deliveryId, CancellationToken ct = default)
        => await PostVoidAsync($"messages/{deliveryId}/read", "Failed to mark message as read.", ct);

    public async Task MarkAllMessagesReadAsync(CancellationToken ct = default)
        => await PostVoidAsync("messages/read-all", "Failed to mark all messages as read.", ct);

    public async Task<int> SendAdminMessageAsync(string title, string content, bool isBroadcast, IReadOnlyList<int>? recipientUserIds = null, DateTimeOffset? expiresAt = null, CancellationToken ct = default)
    {
        var result = await PostFormAsync<SendAdminMessageResponseData>("messages/admin/send", new
        {
            title, content,
            is_broadcast = isBroadcast,
            recipient_user_ids = recipientUserIds,
            expires_at = expiresAt?.ToString("O"),
        }, "Failed to send admin message.", ct);
        return result.DeliveryCount;
    }

    private async Task<TData> GetAuthedAsync<TData>(HttpMethod method, string uri, string fallbackError, CancellationToken ct) where TData : class
    {
        using var request = await CreateAuthedRequestAsync(method, uri, ct);
        using var response = await Http.SendAsync(request, ct);
        var envelope = await ReadEnvelopeAsync<TData>(response, ct);
        if (!response.IsSuccessStatusCode || envelope?.Success != true || envelope.Data is null)
            throw await CreateApiExceptionAsync(response.StatusCode, envelope?.Error?.Message ?? fallbackError, envelope?.Error?.Code, ct);
        return envelope.Data;
    }

    private async Task PostVoidAsync(string uri, string fallbackError, CancellationToken ct)
    {
        using var authRequest = await CreateAuthedRequestAsync(HttpMethod.Post, uri, ct);
        await SendWithoutDataAsync(authRequest, fallbackError, ct);
    }

    private async Task<TData> PostFormAsync<TData>(string uri, object payload, string fallbackError, CancellationToken ct) where TData : class
    {
        using var authRequest = await CreateAuthedRequestAsync(HttpMethod.Post, uri, ct);
        authRequest.Content = JsonContent.Create(payload, options: JsonOptions);
        using var response = await Http.SendAsync(authRequest, ct);
        var envelope = await ReadEnvelopeAsync<TData>(response, ct);
        if (!response.IsSuccessStatusCode || envelope?.Success != true || envelope.Data is null)
            throw await CreateApiExceptionAsync(response.StatusCode, envelope?.Error?.Message ?? fallbackError, envelope?.Error?.Code, ct);
        return envelope.Data;
    }

}
